### ISSUE TITLE
ESM plugin support is very limited, chokes on modern JavaScript syntax

### ISSUE THREAD

**Issue by lpsinger** [architect/architect#1401]
**Describe the issue**
The ESM plugin support is very limited and chokes on a lot of modern JavaScript syntax.

This is probably because it uses the [esm](https://www.npmjs.com/package/esm) package rather than relying on Node.js' built-in ESM support.

**Steps to reproduce**
Create an ESM plugin and add a statement that uses [optional chain gin with bracket notation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#optional_chaining_with_expressions), such as:

```js
export const set = {
  env() {
    let xyzzy = { plugh: 'baf' }
    let baz = xyzzy?.['plugh']  // optional changing with bracket notation
    return { FOOBAR: 'BAT' }
  },
}
```

**Expected behavior**
Plugin run should run with no errors.

**Screenshots**
Running arc sandbox produces an error like this:

```
sandbox failed! Plugin errors:
- SyntaxError: Invalid or unexpected token
- Unable to load plugin 'sandboxOidcIdp': Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at Function.getPluginModules [as plugins] (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/inventory/src/config/pragmas/plugins.js:89:16)
    at architectInventory (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/inventory/src/index.js:67:38)
    at main (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:72:29)
    at /Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:91:11
    at Object.<anonymous> (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:92:5)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
Error: Plugin errors:
- SyntaxError: Invalid or unexpected token
- Unable to load plugin 'sandboxOidcIdp': Cannot convert undefined or null to object
    at Function.entries (<anonymous>)
    at Function.getPluginModules [as plugins] (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/inventory/src/config/pragmas/plugins.js:89:16)
    at architectInventory (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/inventory/src/index.js:67:38)
    at main (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:72:29)
    at /Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:91:11
    at Object.<anonymous> (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:92:5)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
    at format (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/inventory/src/lib/error-fmt.js:15:13)
    at architectInventory (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/inventory/src/index.js:74:14)
    at main (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:72:29)
    at /Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:91:11
    at Object.<anonymous> (/Users/lpsinger/src/gcn.nasa.gov/node_modules/@architect/architect/src/index.js:92:5)
    at Module._compile (node:internal/modules/cjs/loader:1218:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
ERROR: "dev:arc" exited with 1.1
```

**Desktop**
Please complete the following information (if appropriate):
 - OS: macOS
 - Browser N/A
 - Version N/A

**Mobile**
N/A

**Additional context**
Add any other context or notes about the problem here

**Comment by ryanblock** [comment]
Oy, that’s a bummer. I suspect the esm folks may not continue to maintain that library, either. Not sure what the workaround here is except to avoid the latest syntax until we have the time and resources to fully port architect to ESM, which is not a small lift. Open to ideas!

**Comment by ryanblock** [comment]
Having found this thread I am now convinced that adopting esm was a big mistake on my part: https://github.com/standard-things/esm/issues/866

I thought that project could be relied upon; unfortunately that’s not the case. I thought I did enough research into the state of that project, but I was wrong, and I’ll own that mistake. Sorry for the trouble.

There exists a working but unpublished fork that accepts ES2020 syntax (including nullish coalescing), but I need to consider whether it’s better for us to own and maintain that fork vs doing the nontrivial amount of necessary to port Architect to ESM.

If nothing else, now we know the appetite for authoring Arc plugins in ESM is there, so that’s good!

**Comment by ryanblock** [comment]
Followup: I've begun to refactor out `esm` in favor of Node's builtin `await import`, which should resolve this issue. Its module resolution behavior is, of course, different from `esm`'s, and being that it's using "actual" `import` and not shimming `require` to "import" modules, some tests need a massive refactor. I'm guessing this will ship to production this week and we can put this one to bed.

**Comment by ryanblock** [comment]
https://github.com/architect/inventory/pull/54

**Comment by ryanblock** [comment]
Fix released in 10.9.2!
